File: android/SceneWorld001/app/src/main/java/z3d/display/particle/facet/Display3DFacetParticle.java
package z3d.display.particle.facet;

import android.util.Log;

import java.util.List;

import z3d.base.Camera3D;
import z3d.base.ObjData;
import z3d.base.TexTuresBackFun;
import z3d.core.Context3D;
import z3d.display.particle.Display3DParticle;
import z3d.display.particle.ball.ParticleBallData;
import z3d.display.particle.ball.ParticleBallGpuData;
import z3d.filemodel.TextureManager;
import z3d.material.DynamicTexItem;
import z3d.material.TexItem;
import z3d.material.TextureRes;
import z3d.vo.Vector3D;

public class Display3DFacetParticle  extends Display3DParticle {


    private static final String TAG ="Display3DFacetParticle" ;

    @Override
    public void update() {
        super.update();
    }
    public Display3DFacetParticle(){
        super();
    }

    @Override
    public void setVa() {
        super.setVa();
        Context3D ctx=this.scene3d.context3D;
        ObjData objData= this.faceData().objData;
        ctx.setVa(this.shader3D,"v3Position",3,objData.vertexBuffer);
        ctx.setVa(this.shader3D,"v2TexCoord",2,objData.uvBuffer);
        ctx.drawCall(objData.indexBuffer,objData.treNum);
    }
    @Override
    public void setVc() {
        super.setVc();
        Context3D ctx=this.scene3d.context3D;
        this.setViewCamModeMatr3d();
        this.updateRotaionMatrix();
        ctx.setVcMatrix4fv(this.shader3D,"rotMatrix",this.rotationMatrix3D.m);
    }
    private void updateRotaionMatrix() {
        Camera3D cam=this.scene3d.camera3D;
        this.rotationMatrix3D.identity();
        if (this.data._watchEye) {
            if (!this.faceData()._locky && !this.faceData()._lockx) {
                this.inverBind();
            }
            if (!this.faceData()._locky) {
                this.rotationMatrix3D.prependRotation(-cam.rotationY, Vector3D.Y_AXIS);
            }
            if (!this.faceData()._lockx) {
                this.rotationMatrix3D.prependRotation(-cam.rotationX, Vector3D.X_AXIS);
            }
        }
        if (this.data._isZiZhuan) {
           // this.timeline.applySelfRotation(this.rotationMatrix3D,this.data._ziZhuanAngly);
        }
    }

    private void inverBind() {
    }
    public ParticleFacetData faceData(){
        return (ParticleFacetData) this.data;
    }


}

File: listios/StorageStarted/pan3d/display/particle/face/Display3DFacetParticle.m
//
//  Display3DFacetParticle.m
//  iosgl
//
//  Created by zhao on 29/3/2020.
//  Copyright © 2020 zhao. All rights reserved.
//

#import "Display3DFacetParticle.h"
#import "Scene_data.h"
#import "Context3D.h"
#import "Scene3D.h"
#import "ObjData.h"
#import "Vector2D.h"
#import "ParticleFacetData.h"
#import "Display3DFacetShader.h"
#import "ParticleMetalType.h"
 

@interface Display3DFacetParticle ()
@property (nonatomic, strong)  Vector2D*   uvMove;
//@property (nonatomic,strong) MtkBaseDis* mtkBaseLine;
//@property (nonatomic,strong) Display3DFacetShader* mtkBaseLineShader;
@end

 
@implementation Display3DFacetParticle

- (instancetype)init:value
{
    self = [super init:value];
    if (self) {
        self.uvMove=[[Vector2D alloc] init];
 
    }
    return self;
}
 
- (void)update;
{
 
    [super update];
  
 
    
    
    
}
 
 
- (void)setVc;
{
      [self setViewCamModeMatr3d];
      [self updateRotaionMatrix];
      [self updateUV];
    
    
    /*
      Context3D *ctx=self.scene3D.context3D;
      [ctx setVcMatrix4fv:self.shader3d name:"rotMatrix" data:self.rotationMatrix3D.m];
      [ctx setVcUniform2f:self.shader3d name:"uvMove" x:self.uvMove.x y:self.uvMove.y];
    */
    id<MTLRenderCommandEncoder> renderEncoder=self.scene3D.context3D.renderEncoder;
   
    
    ParticleBaseFloat2v uvMoveVc = { (vector_float2){self.uvMove.x,self.uvMove.y }  };
   [renderEncoder setVertexBytes:&uvMoveVc
                          length:sizeof(uvMoveVc)
                         atIndex: 3];
    
    NSLog(@"%f----%f",self.uvMove.x,self.uvMove.y);
}
-(void)updateUV;
{
  
    Display3DFacetParticle* this=self;
    int currentFrame=(int)(this._time/[Scene_data default].frameTime );
    currentFrame = currentFrame > this.facetdata._maxAnimTime ? this.facetdata._maxAnimTime : currentFrame;
    currentFrame=  (int)( currentFrame /this.data._animInterval)%(int)(this.data._animLine * this.data._animRow);
    self.uvMove.x=( (currentFrame % (int)this.data._animLine)/this.data._animLine)+(this._time / [Scene_data default].frameTime* this.data._uSpeed);
    self.uvMove.y=(int)(currentFrame / this.data._animLine)/this.data._animRow+this._time / [Scene_data default].frameTime* this.data._vSpeed;
 
  
}
- (void)setVa;
{
  
   
   id<MTLRenderCommandEncoder> renderEncoder=self.scene3D.context3D.renderEncoder;
   
   
    [renderEncoder setVertexBuffer: self.facetdata.objData.mtkvertices
                            offset:0
                           atIndex:0];
    [renderEncoder setVertexBuffer: self.facetdata.objData.mtkuvs
                            offset:0
                           atIndex:1];
    
    
 
    
   [renderEncoder drawIndexedPrimitives:MTLPrimitiveTypeTriangle
                             indexCount: self.facetdata.objData.mtkindexCount
                              indexType:MTLIndexTypeUInt32
                            indexBuffer: self.facetdata.objData.mtkindexs
                      indexBufferOffset:0];
    
}
- (void)resetVa;
{
    Context3D *ctx=self.scene3D.context3D;
    [ctx clearVa:0];
    [ctx clearVa:1];
 
}
-(void)setViewCamModeMatr3d;
{
 
    Camera3D* cam3D=self.scene3D.camera3D;
    
 
    id<MTLRenderCommandEncoder> renderEncoder=self.scene3D.context3D.renderEncoder;
    
    ParticleMetalMatrixData matrixList = {[cam3D.viewMatrix getMatrixFloat4x4], [cam3D.camMatrix3D getMatrixFloat4x4], [self.modeMatrix getMatrixFloat4x4], [self.rotationMatrix3D getMatrixFloat4x4]};

 
    
   [renderEncoder setVertexBytes:&matrixList
                          length:sizeof(matrixList)
                         atIndex:2];
    

}
-(void)updateRotaionMatrix;
{
    Camera3D *cam=self.scene3D.camera3D;
    Display3DFacetParticle* this=self;
    [this.rotationMatrix3D identity];
    if (this.data._watchEye) {
        if (!this.facetdata._locky && !this.facetdata._lockx) {
            [this inverBind];
        }
        if (!this.facetdata._locky) {
            [this.rotationMatrix3D prependRotation:-cam.rotationY axis:Vector3D.Y_AXIS];
        }
        if (!this.facetdata._lockx) {
            [this.rotationMatrix3D prependRotation:-cam.rotationX axis:Vector3D.X_AXIS];
        }
        
     
    }
    if (this.data._isZiZhuan) {
        // this.timeline.applySelfRotation(this._rotationMatrix, this.data._ziZhuanAngly);
    }
}
-(ParticleFacetData*)facetdata;
{
    return (ParticleFacetData*)self.data;
}
@end
